[feat] Add page structure explorer

bharatkashyap opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->

- Closes #490 

https://github.com/mui/mui-toolpad/assets/19550456/a21773a9-4745-41b9-ba2a-6ccff05dffd9


- [x] I've read and followed the [contributing guide](https://github.com/mui/mui-toolpad/blob/master/CONTRIBUTING.md#sending-a-pull-request) on how to create great pull requests.
- [x] I've updated the relevant documentation for any new or updated feature
- [x] I've linked relevant GitHub issue with "Closes #<issue id>"
- [x] I've added a visual demonstration under the form of a screenshot or video

Janpot commented on test/integration/backend-basic/fixture/toolpad/pages/serialization/page.yml:
```diff
@@ -46,3 +46,27 @@ spec:
             value:
               $$jsExpression: |
                 `Invalid error: ${invalidError.error?.message}`
+    - component: PageRow
```
Accidentally committed?

Janpot commented:
Something seems to go wrong with the hover state when renaming:

https://github.com/mui/mui-toolpad/assets/2109932/b832b954-2d69-4830-ad98-e2e31addeed2

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/PageEditor/RenderPanel/NodeHud.tsx:
```diff
@@ -167,6 +169,7 @@ export default function NodeHud({
   onDuplicate,
   isOutlineVisible = false,
   isHoverable = true,
+  isHovered = false,
```
Since the dashed border doesn't really just mean the element is "hovered" anymore we might want to give this prop a different name just to avoid confusion. Maybe `isPreviewed`? Not sure, also not a super big deal.

apedroferreira replied:
Ok, I think it's fine to keep it.

apedroferreira reviewed (changes requested):
Looks awesome and works great! Just noticed a couple of issues:
- When I drag-and-drop to create a column the whole row that contains it collapses in the panel
- It's not possible yet to press backspace to delete after selecting a row or column in the explorer, but I feel like it would be a really useful feature

prakhargupta1 commented:
> * Do you think it's time we start using `TreeView` for this UI? I feel like it would create denser UI.
> * Should we add some sort of section title above that says "explorer"?

Looks good! +1 for both of these suggestions:

bharatkashyap commented:
I've updated to address most of the feedback here:

https://github.com/mui/mui-toolpad/assets/19550456/b7fd5446-5faa-4edd-8810-bc5e62d06387

@apedroferreira I've added an implementation of deleting via the explorer which plays the most nicely with undoing, does this make sense to you are are there any other ways of achieving this?

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/StructureExplorer/index.tsx:
```diff
@@ -187,40 +213,74 @@ export default function PageStructureExplorer() {
+            return draft;
+            // return normalizePageRowColumnSizes(draft);
+          },
+          currentView.kind === 'page'
```
I wonder if we can get rid of these `currentView.kind === 'page'` as I think these actions can only occur in a page view? But I may have forgotten why I added these checks in the other similar deletes...

bharatkashyap replied:
I think it's for the types, in which case I can probably replace these conditionals with 
```...(currentView as Extract<DomView, { kind: 'page' }>),```

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/StructureExplorer/index.tsx:
```diff
@@ -187,40 +213,74 @@ export default function PageStructureExplorer() {
+    (event: React.KeyboardEvent<HTMLUListElement>) => {
+      // delete selected node if event.key is Backspace
+      if (event.key === 'Backspace') {
+        appStateApi.update(
```
Seems like the correct logic, except we do may need to normalize the page row column sizes after deleting as done in the other places...
And now that this logic goes beyond just the `RenderOverlay` component, we should move all of this (deleteing node + oprhaned nodes + normalizing column sizes) to somewhere where we can reuse it more agnostically, not sure if that's easy or not.

bharatkashyap replied:
> And now that this logic goes beyond just the RenderOverlay component, we should move all of this (deleteing node + deleting orphaned nodes + normalizing column sizes) to somewhere where we can reuse it more 
I've added some changes to try and accomplish this, have a look

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/StructureExplorer/index.tsx:
```diff
@@ -187,40 +213,74 @@ export default function PageStructureExplorer() {
+        onNodeSelect={handleNodeSelect}
+        onNodeFocus={handleNodeFocus}
+        onNodeToggle={handleNodeToggle}
+        onKeyDown={handleNodeDelete}
```
I would separate `onKeyDown` to something like `handleKeyDown`, and there is backspace is pressed call a reusable delete function. 
That way for example if we allow deleting in the explorer by clicking the button we just have to use the same function that is already there.

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/StructureExplorer/index.tsx:
```diff
@@ -0,0 +1,286 @@
+      if (event.key === 'Backspace') {
+        appStateApi.update(
+          (draft) => {
+            if (!selectedDomNodeId) {
```
Can you check for if there is a selected node outside of `appStateApi.update`? That way we don't add a new entry to the undo stack, I think.

apedroferreira reviewed (changes requested):
Looks good, just added comments on some minor things but still everything is working great!

prakhargupta1 commented:
Just played with it, gives so much clarity about the page hierarchy now. Awesome UX! 🙌

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/PageEditor/RenderPanel/RenderOverlay.tsx:
```diff
@@ -268,6 +268,51 @@ export function deleteOrphanedLayoutNodes(
   return draftDom;
 }
 
+const rowColumnCounts: Record<NodeId, number> = {};
```
Looks good if it works! We could probably reset this variable if the page changes to keep it from growing too big, but no need to do it yet probably.

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/utils/deleteNode.ts:
```diff
@@ -0,0 +1,31 @@
+) {
+  appStateApi.update(
+    (draft) => {
+      const toRemove = appDom.getNode(draft, nodeId);
```
What if we kept just the part inside `appStateApi.update` as this reusable function, and so we didn't need to pass `appStateApi` and `currentView` as arguments, it would just be a function that transforms the DOM? 
I feel like it would be clearer that way and depend less on context, even if we had to repeat the other parts when we called the function.

apedroferreira commented on packages/toolpad-app/src/toolpad/AppEditor/utils/normalizePageRowColumnSizes.ts:
```diff
@@ -0,0 +1,54 @@
+const rowColumnCounts: Record<NodeId, number> = {};
+
+export default function normalizePageRowColumnSizes(
+  draftDom: appDom.AppDom,
```
Perhaps we can remove the word `draft` from this function as the usage is more general now, i guess it could be just `dom` instead of `draftDom` for example?

apedroferreira reviewed (approved):
Awesome work Bharat, looks really great! Thanks a lot for addressing these!
I still had some thoughts and possible suggestions but I don't think any of them should be that blocking, it's very minor stuff, so I will approve already.
Maybe a small refactor of `deleteNode` would be the most important one to me!

bharatkashyap commented:
@apedroferreira Resolving the comments related to the common delete function extraction since we agreed on a different approach involving refactoring the existing delete

apedroferreira commented:
> @apedroferreira Resolving the comments related to the common delete function extraction since we agreed on a different approach involving refactoring the existing delete

Ok, sounds good!
